File: 008/008.cpp
class Solution
{
public:
    int myAtoi(string s)
    {
        int n = s.length();
        int i = 0;
        while (i < n && s[i] == ' ')
        {
            i++;
        }

        if (i == n || !((s[i] >= '0' && s[i] <= '9') || s[i] == '+' || s[i] == '-'))
        {
            return 0;
        }

        bool isNegative = false;
        if (s[i] == '+' || s[i] == '-')
        {
            if (s[i] == '-')
            {
                isNegative = true;
            }
            i++;
            if (i == n)
            {
                return 0;
            }
        }

        long result = 0;
        while (i < n && s[i] >= '0' && s[i] <= '9')
        {
            result = 10 * result + (s[i] - '0');
            if (isNegative && result > pow(2, 31))
            {
                return pow(2, 31);
            }
            else if (!isNegative && result >= pow(2, 31))
            {
                return pow(2, 31) - 1;
            }
            i++;
        }

        if (isNegative)
        {
            return int(-result);
        }
        else
        {
            return int(result);
        }
    }
};

File: 008/008.py
class Solution:
    def myAtoi(self, s):

        ## Input:
        # s: str
        ## Output:
        # int

        n = len(s)
        i = 0
        while (i < n and s[i] == ' '):
            i += 1

        if (i == n or not ((s[i] >= '0' and s[i] <= '9') or s[i] == '+' or s[i] == '-')):
            return 0
        
        isNegative = False
        if (s[i] == '+' or s[i] == '-'):
            if (s[i] == '-'):
                isNegative = True
            i += 1
            if (i == n):
                return 0
        
        result = 0
        while (i < n and s[i] >= '0' and s[i] <= '9'):
            result = 10 * result + int(s[i])
            if (isNegative and result > 2 ** 31):
                return -(2 ** 31)
            elif (not isNegative and result >= 2 ** 31):
                return 2 ** 31 - 1
            i += 1
        
        if (isNegative):
            return -result
        else:
            return result